src/libmcrx.c

```diff
@@ -1109,16 +1108,39 @@ MCRX_EXPORT enum mcrx_error_code mcrx_mnat_ctx_new(
  *
  **/
 MCRX_EXPORT enum mcrx_error_code mcrx_mnat_ctx_apply(struct mcrx_ctx *ctx,
-    struct mcrx_mnat_ctx *mnatctx) {
+    struct mcrx_mnatmap *mnatmap) {
   if (ctx == NULL) {
     return MCRX_ERR_NULLARG;
   }
-  if (ctx->mnat_ctx == NULL) {
+  if (ctx->mnat_map == NULL) {
     // no mnat running time mcrx context yet
-    mcrx_mnat_ctx_clone(mnatctx, &ctx->mnat_ctx);
+    mcrx_mnat_ctx_clone(mnatmap, &ctx->mnat_map);
+    struct mcrx_subscription *sub;
+    LIST_FOREACH(sub, &ctx->subs_head, sub_entries)
+    {
+      // find new entry based on the new mnat context
+      struct mcrx_mnat_entry *new_entry =
+          mcrx_mnat_ctx_find_entry_from_subscription(sub, ctx->mnat_map);
+      if (sub->joined) {
+        if (new_entry == NULL
+            || !mcrx_mnat_ctx_entry_local_equal(sub->mnat_entry, new_entry)) {
```

Comment:
I have some concerns about the logic inside this loop (and the corresponding else's loop).
I think we should have some comments here that describe the invariants and states we're maintaining, because I'm finding it a little hard to think through in my head.

From the logic here, I think the current intent seems to be:
- if a sub is not joined, its mnat_entry is null
- if a sub is joined and ctx->mnat_map is null, the sub's mnat_entry is null
- if a sub is joined and ctx->mnat_map is non-null, the sub's mnat_entry points to an entry owned by ctx->mnat_map
- if a sub has a non-null mnat_entry, it's either resolved or unresolved according to the local (s,g) in its mnat_entry.
  - If unresolved, there's no underlying native socket
  - if resolved the underlying native socket is attempted and errors are reported if unsuccessful.
- a sub is joined if & only if the underlying native socket is operating without error
- ctx->live_subs holds the count of currently joined subs

I'm not sure all these invariants are maintained (and I'm also not sure it's the correct list of relevant invariants and states here).

I think we should probably meet to walk through this logic, and cross-check it with what happens when a user requests joining of a sub when there's a mnatmap.

If we have a list we think is correct, we maybe should also add a way to check the invariants (maybe empty in release?),and call it at the beginning and end of this and a few of the other more complex functions.

Reply:
From discussion:
- 2nd one is not true: mnat_entry should be non-null even for non-joined subs with a corresponding entry in the map.
  - this means we need to edit the subs that are not joined also, to make the mnat_entry point to the right thing.

Reply:
I think we need a new state here:
- user attempted to join, but mnat is active and this global (S,G) is unresolved

In this state, sub's mnat_entry should be non-null and unresolved.

Reply:
MCRX_EXPORT enum mcrx_error_code mcrx_subscription_join(
    struct mcrx_subscription* sub) {
  if (sub && sub->ctx && sub->ctx->mnat_map) {
    sub->mnat_entry = mcrx_mnat_ctx_find_entry_from_subscription(sub, sub->ctx->mnat_map);
    if (sub->mnat_entry == NULL || mcrx_mnat_ctx_entry_unresolved(sub->mnat_entry)) {
      // if we can not find a mnap entry or the entry is unresolved for local address
      info(sub->ctx, "can not find resolved mnat entry for subscription %p\n", (void *)sub);
      return MCRX_ERR_UNSUPPORTED;
    }
  }

@GrumpyOldTroll do we need to take care of case 1) no mnap entry 2) mnap entry local is not resolved and return different error codes. We can introduce a new API to get all unresolved mnap entries for subs

Reply:
I think the right behavior if there is a subscription join and mnat is active for the ctx is to create a new unresolved mnat_entry and attach it to the subscription, so subsequent clones of the mnatmap in the ctx would contain the unresolved entry.

We also probably should remove entries from the context's mnatmap for subscriptions that are no longer in join_desired (when they get a leave or get unref'd the last time).

There's a chance for some use cases (especially ones like ABR that know they need to switch between a known small set of streams that will have some joined and others not) that the right behavior is actually to decouple this from the join/leave, so that all subscriptions in the ctx's list of subscriptions would always have a mnat_entry, which would be unresolved.  I'm not sure which way is better, now that I think of it.  But I think we at least need the entries to always be there (even if it's not yet resolved) when a join is desired.